### DISCUSSION TITLE
Advice needed (by another open-source AirPlay application)  about AirPlay volume slider implementation

### DISCUSSION THREAD

**Discussion by fduncanh** [mikebrady/shairport-sync#1773] [Q&A]
(this is from the maintainer of a different open-source AirPlay application , [UxPlay](http://github.com/FDH2/UxPlay), an AppleTV emulator)

We are having a [discussion](http://github.com/FDH2/UxPlay/pull/250)   about  what the "correct" implementaion of Apple's audio volume  is.  The "unofficial" AirPlay specification alleges that the values "volume" ( a floating point number in range [-30:0] inclusive) are (power?) decibels

 Conversion to a linear scale (UxPlay uses GStreamer for audio after RAOP decryption)  would then be (in C) appear to be
```
linear = powf(10, volume /10)
```
(probably set to zero for volume -30 or below,, and to 1 for 0 and above)

However this is not entirely satisfactory, and there seem to be issues with AirPlay volume in proprietary non-Apple 
implememtations too.

* what does shairport-sync do?

* does anyone here have any insights into Apple AirPlay / AirPlay2 volume-control quirks?

**Comment by mikebrady** [comment]
Thanks for the message. There is another slight wrinkle in this -- the range is -30.0 to 0.0 alright, but additionally the value -144.0 means "mute".

The -30.0 to 0.0 range does suggest some kind of logarithmic range, but AFAIK there is no documentary evidence to support this.

Shairport Sync controls hardware or software mixers that have decibel-denominated settings -- it doesn't even try to control mixers without dB-denominated settings.

There is a wide range of mixer attenuation ranges out there, from about 30 dB on cheap USB DACs to well over 100 dB on others. In reality, a range of 70 to 80 dB seems necessary. Shairport Sync allows you to specify the range you wish to use -- it can also augment a limited hardware mixer range with an additional software mixer. (It can also be set to ignore volume control information completely.)

Shairport Sync has three optional transfer functions between the AirPlay range and the dB ranges of the mixers:

1. The default transfer function is piecewise linear between AirPlay volume and dB mixer volume but with slight changes at the low and high ends of the range to make the volume control a bit more "natural" and responsive.
2. The "flat" transfer function is linear between AirPlay and the dB range of the mixer.
3. The (new) "dasl-tapered" transfer function has the property that moving the AirPlay volume slider through half of its range approximately "halves" the volume. Thus, going from 0.0 to -15.0 reduces the volume by 10dB; from -15.0 to -22.5 drops it by a further 10 db, and so on.

There is quite a bit of information on the Internet about the properties of "ideal" volume controls.

**Comment by fduncanh** [comment]
Thanks!

very helpful!
"dasl-tapered" looks interesting.

**Comment by soul916** [comment]
@mikebrady Thank you for your detailed explanation. I have another question about "the value -144.0 means mute.", I can't set the volume to this value with my IOS or MAC OS device. Under what situation will this value appear?

**Reply by mikebrady** [reply]
The -144.0 is sent by the player when the slider is right down at the bottom left. It jumps from something just above -30.0 to -144.0. Just tried it with iOS (latest version) Music app on AirPlay 2.

**Reply by soul916** [reply]
Thank you for the answer!
